navis/nbl/nblast_funcs.py

```diff
@@ -124,15 +147,24 @@ def __init__(self, use_alpha=False, normalized=True, smat='auto', progress=True)
         self.normalized = normalized
         self.progress = progress
 
-        if smat == 'auto':
+        if smat is None:
+            smat = operator.mul
+        elif smat == 'auto':
             if self.use_alpha:
-                smat = pd.read_csv(f'{smat_path}/smat_alpha_fcwb.csv',
-                                   index_col=0)
+                smat = smat_path / 'smat_alpha_fcwb.csv'
             else:
-                smat = pd.read_csv(f'{smat_path}/smat_fcwb.csv',
-                                   index_col=0)
+                smat = smat_path / 'smat_fcwb.csv'
+
+        if isinstance(smat, (str, os.PathLike)):
+            smat = pd.read_csv(smat, index_col=0)
+
+        if isinstance(smat, pd.DataFrame):
+            smat = Lookup2d.from_dataframe(smat)
+
+        if not callable(smat):
```

Comment:
I'm a big fan of very verbose error messages (for dummies like myself): in case somebody passed a `callable` as `smat`, it might be a good idea to test if it's valid: i.e. it accepts two (N, ) numpy arrays and returns a single (N, 3) array.

Reply:
Completely agree on failing quickly and explicitly, and this would work fine for the moment. We'd need to consider how to keep that strictness in a possible arbitrary N-dimensional score function future, where we don't necessarily know how many arguments it should take (although we could use introspection), or what the domain of the function is (if it's only defined within certain values, we don't want to refuse to use it just because we don't know what those values are).

But a fair bit would have to change elsewhere in the code for us to even get to that future, so we can cross that bridge when we come to it.

Reply:
Sweet! In theory, the `NBlaster` class (and any variants of it like `SynBlaster`) should "know" what they expect from the scoring function. Maybe it would be most convenient to make it a class method like `validate_score_func(self)`?

Reply:
Also: I feel like I'm making you jump through too many hoops here. I'm perfectly happy with what's there right now and would not mind just merging if your commits already cover your use case.

Reply:
There is now a functional function-checker, which is factored into its own function which takes as an argument the number of arguments the checked function should take. Phew, that's a mouthful. Anyway, it would be easy enough to check that at the class level in future.